demos/source/aws_iot_demo_defender.c

```diff
@@ -0,0 +1,226 @@
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/* Standard includes. */
+#include <stdio.h>
+#include <string.h>
+
+/* Demo configuration includes. */
+
+/* Demo logging include. */
+#include "iot_demo_logging.h"
+
+/* Platform includes for demo. */
+#include "platform/iot_clock.h"
+#include "platform/iot_network.h"
+
+/* Defender includes. */
+#include "aws_iot_defender.h"
+
+/* Includes for initialization. */
+#include "iot_mqtt.h"
+
+/**
+ * @brief Runs the defender demo.
+ *
+ * @return AWS_IOT_DEFENDER_SUCCESS on success, otherwise an error code indicating
+ *         the cause of error.
+ */
+static AwsIotDefenderError_t _defenderDemo( const char *pIdentifier,
+                                            void *pNetworkServerInfo,
+                                            void *pNetworkCredentialInfo,
+                                            const IotNetworkInterface_t *pNetworkInterface );
+
+/**
+ * @brief Starts the defender agent.
+ *
+ * @return AWS_IOT_DEFENDER_SUCCESS on success, otherwise an error code indicating
+ *         the cause of error.
+ */
+static AwsIotDefenderError_t _startDefender( const char *pIdentifier,
+                                             void *pNetworkServerInfo,
+                                             void *pNetworkCredentialInfo,
+                                             const IotNetworkInterface_t *pNetworkInterface );
+
+/**
+ * @brief Callback used to get notification of defender's events.
+ */
+static void _defenderCallback( void *param1, AwsIotDefenderCallbackInfo_t *const pCallbackInfo );
+
+/*-----------------------------------------------------------*/
+
+int RunDefenderDemo( bool awsIotMqttMode,
+                     const char *pIdentifier,
+                     void *pNetworkServerInfo,
+                     void *pNetworkCredentialInfo,
+                     const IotNetworkInterface_t *pNetworkInterface )
+{
+    int status = EXIT_FAILURE;
+    IotMqttError_t mqttInitStatus;
+    AwsIotDefenderError_t defenderResult;
+
+    /* Unused parameters */
+    ( void ) awsIotMqttMode;
+
+    /* Initialize Metrics */
+    IotMetrics_Init();
+
+    /* Initialize the MQTT library. */
+    mqttInitStatus = IotMqtt_Init();
+
+    if( pIdentifier == NULL || pIdentifier[ 0 ] == '\0' )
+    {
+        IotLogError( "Empty Identifier Use." );
+    }
+    else if( mqttInitStatus == IOT_MQTT_SUCCESS )
+    {
+        /* If the MQTT initialization was successful, run the demo. */
+        defenderResult = _defenderDemo( pIdentifier,
+					pNetworkServerInfo,
+					pNetworkCredentialInfo,
+					pNetworkInterface );
+
+        if( defenderResult == AWS_IOT_DEFENDER_SUCCESS )
+        {
+            status = EXIT_SUCCESS;
+        }
+    }
+
```

Comment:
At the end of the demo, there should be a call to `IotMqtt_Cleanup` and `IotMetrics_Cleanup`.